File: src/javascript/algorithms/binary/3.countingBits/countingBits.js
// Dynamic programming:- TC:O(n) SC:O(n)

function countingBits1(num) {
    let ans = new Array(num+1).fill(0);
    let offset = 1; 
    for(let i=1; i< num+1; i++) {
        if(offset * 2 === i) {
            offset = i;
        }
        ans[i] = 1 +ans[i-offset];
    }
    return ans;
}

function countingBits2(num) {
    let ans = new Array(num+1).fill(0);
    for(let i=1; i< num+1; i++) {
        ans[i] = ans[i >> 1] + (i & 1);
    }
    return ans;
}

let num1 = 3;
let num2 = 6;

console.log(countingBits1(num1));
console.log(countingBits1(num2));

console.log(countingBits2(num1));
console.log(countingBits2(num2));

File: src/python/algorithms/binary/3.countingBits/counting_bits.py
"""
Counting Bits

Given an integer n, return an array ans of length n + 1 such that for each i (0 <= i <= n),
ans[i] is the number of 1's in the binary representation of i.

Time Complexity: O(n)
Space Complexity: O(n)
"""


def count_bits(n):
    """
    Dynamic programming approach.
    TC: O(n), SC: O(n)
    """
    dp = [0] * (n + 1)
    offset = 1
    
    for i in range(1, n + 1):
        if offset * 2 == i:
            offset = i
        dp[i] = 1 + dp[i - offset]
    
    return dp


def count_bits_alternative(n):
    """
    Using bit manipulation.
    TC: O(n), SC: O(n)
    """
    dp = [0] * (n + 1)
    
    for i in range(1, n + 1):
        # dp[i] = dp[i >> 1] + (i & 1)
        dp[i] = dp[i // 2] + (i % 2)
    
    return dp


# Test cases
if __name__ == "__main__":
    test_cases = [2, 5, 8]
    
    for n in test_cases:
        print(f"n = {n}")
        print(f"Counting bits: {count_bits(n)}")
        print()
